File: python/minesweeper/minesweeper.py
"""Minesweeper."""

ADJACENT = [complex(x, y) for x in (-1, 0, 1) for y in (-1, 0, 1) if x or y]


def annotate(minefield: list[str]) -> list[str]:
    """Annotate a minefield."""
    height = len(minefield)
    width = len(minefield[0]) if minefield else 0

    if not all(len(row) == width for row in minefield):
        raise ValueError("The board is invalid with current input.")

    mines = set()
    for y, line in enumerate(minefield):
        for x, val in enumerate(line):
            if val == "*":
                mines.add(complex(x, y))
            elif val != " ":
                raise ValueError("The board is invalid with current input.")

    def cell_value(point: complex) -> str:
        """Return the value for one square."""
        if point in mines:
            return "*"
        count = sum(point + offset in mines for offset in ADJACENT)
        return str(count) if count else " "

    return [
        "".join(cell_value(complex(x, y)) for x in range(width))
        for y in range(height)
    ]

File: go/minesweeper/minesweeper.go
package minesweeper

import (
	"strconv"
	"strings"
)

type point struct{ x, y int }

// neighbors returns the 8 adjacent points.
func (p point) neighbors() []point {
	neighbors := make([]point, 0, 8)
	for x := -1; x <= 1; x++ {
		for y := -1; y <= 1; y++ {
			if x != 0 || y != 0 {
				neighbors = append(neighbors, point{p.x + x, p.y + y})
			}
		}
	}
	return neighbors

}

type minefield struct {
	mines         map[point]bool
	width, height int
}

func newMinefield(board []string) minefield {
	mines := map[point]bool{}

	for y, line := range board {
		for x, char := range line {
			switch char {
			case '*':
				mines[point{x, y}] = true
			case ' ':
			default:
				panic("Invalid char " + string(char))
			}
		}
	}

	width := 0
	if len(board) > 0 {
		width = len(board[0])
	}

	return minefield{mines, width, len(board)}
}

// cell returns the annotation for a single cell.
func (m minefield) cell(p point) string {
	if m.mines[p] {
		return "*"
	}
	count := 0
	for _, n := range p.neighbors() {
		if m.mines[n] {
			count++
		}
	}
	if count == 0 {
		return " "
	}
	return strconv.Itoa(count)
}

func (m minefield) annotate() []string {
	out := make([]string, m.height)
	for y := 0; y < m.height; y++ {
		var sb strings.Builder
		for x := 0; x < m.width; x++ {
			sb.WriteString(m.cell(point{x, y}))
		}
		out[y] = sb.String()
	}
	return out
}

// Annotate returns an annotated board
func Annotate(board []string) []string {
	return newMinefield(board).annotate()
}
